Add config for returning dummy object in the NRTMv3 response

Justin-APNIC opened the pull request:
Add new configs in IRRD for returning the dummy object in NRTMv3 response

Example config
```
                nrtm_response_dummy_object_class:
                  - person
                  - role
                nrtm_response_dummy_attributes:
                  person: Dummy name for %s
                  role: Dummy role for %s
                  address: Dummy address for %s
                  phone: '+31205354444'
                  e-mail: unread@ripe.net
                  upd-to: unread@ripe.net
                  descr: Dummy description for %s
                nrtm_response_dummy_remarks: |
                  ****************************
                  * THIS OBJECT IS NOT VALID
                  * Please note that all personal data has been removed from this object.
                  * To view the original object, please query the APNIC Database at:
                  * http://www.apnic.net/whois
                  ****************************
```

Example response
```
role:           Dummy role for AT1599-AP
address:        Dummy address for AT1599-AP
country:        ZZ
phone:          +31205354444
e-mail:         unread@ripe.net
admin-c:        TPLA28-AP
tech-c:         TPLA28-AP
nic-hdl:        AT1599-AP
mnt-by:         APNIC-ABUSE
last-modified:  2024-03-14T23:45:51Z
source:         APNIC
remarks:        ****************************
remarks:        * THIS OBJECT IS NOT VALID
remarks:        * Please note that all personal data has been removed from this object.
remarks:        * To view the original object, please query the APNIC Database at:
remarks:        * http://www.apnic.net/whois
remarks:        ****************************
```

Justin-APNIC committed:
Add config for returning dummy object in the NRTMv3 response

Justin-APNIC committed:
lint

Justin-APNIC committed:
Fix failed tests

Justin-APNIC committed:
Lint

Justin-APNIC committed:
Add more tests

Justin-APNIC committed:
Lint

mxsasha commented:
Nice! This will take a bit of time to review, but my first thought is that the logic to generate dummified object text should be separated to make it easy to apply to RPSL full exports and NRTMv4 as well. Probably in the same place as remove_auth_hashes(). Makes tests a bit simpler too.

(Initially I thought about putting it on RPSLObject, but we intentionally don't have that in this situation, just the text from the database.)

Justin-APNIC commented:
> Nice! This will take a bit of time to review, but my first thought is that the logic to generate dummified object text should be separated to make it easy to apply to RPSL full exports and NRTMv4 as well. Probably in the same place as remove_auth_hashes(). Makes tests a bit simpler too.

Yes, the function `dummy_rpsl_object()` does exist in the text.py where the` remove_auth_hashes()` function is. The codes in nrtm_generator are just for retrieving the params that the dummy function needs from the configuration file.

Justin-APNIC committed:
Use the pk stored in the db

Justin-APNIC committed:
Lint

Justin-APNIC commented:
@mxsasha
I have a question regarding the primary key in the RPSL object. It appears that we are converting the primary key to uppercase all the time. Could we add any configurations or functions in the rpsl/parser.py to preserve the original primary key case? I've reviewed the code, and it seems that IRRD parses attributes in different object classes differently. Currently, I don't have a straightforward solution, so I'd appreciate your input on this matter.

The reason I want to keep the original value is that I want to return the original pk instead of the uppercase one in the dummy NRTM response.
For example, I want to have IRT-Youminet-CN in the dummy address instead of IRT-YOUMINET-CN

```
irt:            IRT-Youminet-CN
address:        Dummy address for IRT-YOUMINET-CN
```

mxsasha commented:
> > Nice! This will take a bit of time to review, but my first thought is that the logic to generate dummified object text should be separated to make it easy to apply to RPSL full exports and NRTMv4 as well. Probably in the same place as remove_auth_hashes(). Makes tests a bit simpler too.
> 
> Yes, the function `dummy_rpsl_object()` does exist in the text.py where the` remove_auth_hashes()` function is. The codes in nrtm_generator are just for retrieving the params that the dummy function needs from the configuration file.

What I meant is: to properly support dummy mirroring, dummifying needs to happen in the NRTM v3 generator, the source export (if unfiltered is not set, currently called "remove_auth_hashes" I think), and the NRTMv4 server. With as little duplication as possible. So the call from the NRTMv3 generator should be something like: `text = dummify_object_text(source, object_class, text)`.

> @mxsasha I have a question regarding the primary key in the RPSL object. It appears that we are converting the primary key to uppercase all the time. Could we add any configurations or functions in the rpsl/parser.py to preserve the original primary key case? I've reviewed the code, and it seems that IRRD parses attributes in different object classes differently. Currently, I don't have a straightforward solution, so I'd appreciate your input on this matter.

The only place where we retain the original case for the PK is in the object text, so we'd have to re-parse the whole object. PKs are case insensitive, so the indexed key is normalised. But I don't follow why you need it, your current implementation in text.py seems the most straightforward, without needing the PK separately.

Justin-APNIC committed:
Codes refactoring

Justin-APNIC committed:
Lint and fix test coverage error

Justin-APNIC commented:
> > > Nice! This will take a bit of time to review, but my first thought is that the logic to generate dummified object text should be separated to make it easy to apply to RPSL full exports and NRTMv4 as well. Probably in the same place as remove_auth_hashes(). Makes tests a bit simpler too.
> > 
> > 
> > Yes, the function `dummy_rpsl_object()` does exist in the text.py where the` remove_auth_hashes()` function is. The codes in nrtm_generator are just for retrieving the params that the dummy function needs from the configuration file.
> 
> What I meant is: to properly support dummy mirroring, dummifying needs to happen in the NRTM v3 generator, the source export (if unfiltered is not set, currently called "remove_auth_hashes" I think), and the NRTMv4 server. With as little duplication as possible. So the call from the NRTMv3 generator should be something like: `text = dummify_object_text(source, object_class, text)`.
> 
Ok

> > @mxsasha I have a question regarding the primary key in the RPSL object. It appears that we are converting the primary key to uppercase all the time. Could we add any configurations or functions in the rpsl/parser.py to preserve the original primary key case? I've reviewed the code, and it seems that IRRD parses attributes in different object classes differently. Currently, I don't have a straightforward solution, so I'd appreciate your input on this matter.
> 
> The only place where we retain the original case for the PK is in the object text, so we'd have to re-parse the whole object. PKs are case insensitive, so the indexed key is normalised. But I don't follow why you need it, your current implementation in text.py seems the most straightforward, without needing the PK separately.

My implementation in text.py needs a pk to be provided. As I mentioned above, If my config for the dummy attribute is like 
`address: Dummy address for %s
`
The `%s` will be replaced by the pk. However, the current codes will convert it to the uppercase if I call pk() function on the whole object or retrieve the pk from db directly, so the final dummy attribute will be 
`address:        Dummy address for IRT-YOUMINET-CN` instead of `address:        Dummy address for IRT-Youminet-CN`. 

I can't find an easy way to find the original pk from the whole object.

Justin-APNIC committed:
Add config for keeping original data in NRTM stream for specific source

Justin-APNIC committed:
Fix spelling error and failed test

Justin-APNIC committed:
Add new word to spelling word list

mxsasha commented:
> My implementation in text.py needs a pk to be provided. As I mentioned above, If my config for the dummy attribute is like `address: Dummy address for %s ` The `%s` will be replaced by the pk. However, the current codes will convert it to the uppercase if I call pk() function on the whole object or retrieve the pk from db directly, so the final dummy attribute will be `address: Dummy address for IRT-YOUMINET-CN` instead of `address: Dummy address for IRT-Youminet-CN`.
> 
> I can't find an easy way to find the original pk from the whole object.

There is none. The data simply isn't extracted, as PKs are case insensitive in every other usage, and therefore normalised to upper case during parsing of the initial object. Is it such an obstacle for you? Does the PK need to be repeated in the dummy text? After all, anyone who looks at the object will see the true PK already anyways.

mxsasha commented on irrd/conf/__init__.py:
```diff
@@ -676,6 +678,19 @@ def get_object_class_filter_for_source(source: str) -> Optional[List[str]]:
         return None
 
 
+def get_nrtm_response_dummy_object_class_for_source(source: str) -> Optional[List[str]]:
```
This function is better placed in text.py (or even inlined there) as it's the only place where we use it.

Justin-APNIC replied:
Done, thanks

mxsasha commented on irrd/mirroring/nrtm_generator.py:
```diff
@@ -21,6 +22,7 @@ def generate(
         serial_end_requested: Optional[int],
         database_handler: DatabaseHandler,
         remove_auth_hashes=True,
+        dummify_object_text=True,
```
This specifically relates to the client IP. The actual decision on alteration is based on more parameters later on. So I think this needs a more narrow name, like (note inverse logic): `client_is_dummifying_exempt=False`.

Justin-APNIC replied:
Done, thanks

mxsasha commented on irrd/utils/text.py:
```diff
@@ -97,3 +103,46 @@ def _str_to_camel_case(snake_str: str):
+
+def dummify_object_text(rpsl_text: str, object_class: str, source: str, pk: str):
+    """
+    Modify the value of attributes in an RPSL object.
```
This is not a very descriptive docstring.

Justin-APNIC replied:
Added more comments, thanks.

mxsasha commented on irrd/utils/text.py:
```diff
@@ -97,3 +103,46 @@ def _str_to_camel_case(snake_str: str):
+
+    nrtm_response_dummy_object_class = get_nrtm_response_dummy_object_class_for_source(source)
+    if nrtm_response_dummy_object_class:
+        if object_class in nrtm_response_dummy_object_class:
```
We can simplify a level:
```suggestion
    nrtm_response_dummy_object_class = get_nrtm_response_dummy_object_class_for_source(source, [])
    if object_class in nrtm_response_dummy_object_class:
```

Justin-APNIC replied:
Done, thanks

mxsasha commented on irrd/utils/text.py:
```diff
@@ -97,3 +103,46 @@ def _str_to_camel_case(snake_str: str):
+            if dummy_attributes:
+                if get_setting(f"sources.{source}.nrtm_response_dummy_remarks"):
+                    dummy_remarks = textwrap.indent(
+                        get_setting(f"sources.{source}.nrtm_response_dummy_remarks"), "remarks:".ljust(16)
```
Where does that 16 come from? From somewhere in irrd.rpsl right? Feels like we should make that a constant then.

Justin-APNIC replied:
yes, it comes from RPSL_ATTRIBUTE_TEXT_WIDTH which has been used in different places. Updated to use the constant, thanks

mxsasha commented on irrd/utils/text.py:
```diff
@@ -97,3 +103,46 @@ def _str_to_camel_case(snake_str: str):
+
+                        if line.startswith(f"{key}:"):
+                            format_key = f"{key}:".ljust(RPSL_ATTRIBUTE_TEXT_WIDTH)
+                            if not isinstance(value, str):
```
In which cases is this not a string? And couldn't we just always use `str(value)`?

Justin-APNIC replied:
Yes, it can be simplified. Thanks.

mxsasha commented on irrd/utils/text.py:
```diff
@@ -97,3 +103,46 @@ def _str_to_camel_case(snake_str: str):
+
+                dummyfied_rpsl_object = "\n".join(lines) + "\n"
+
+                if rpsl_text != dummyfied_rpsl_object:
```
Creating dummy_remarks could probably happen inside this if statement. So we don't calculate it if there is no change.

Justin-APNIC replied:
This is how the current codes do. If no change or no remarks configured, we won't add the dummy remarks. Am I misunderstanding something here?

mxsasha reviewed (commented):
Left some notes inline. We also need to extend this to the source export runner, and the NRTM4 server. In those cases, we don't have client IPs, so our only option is to always enable it if configured.

I also want to tweak the docs and settings names a bit, later.

Justin-APNIC committed:
Codes refactoring

Justin-APNIC commented:
> > My implementation in text.py needs a pk to be provided. As I mentioned above, If my config for the dummy attribute is like `address: Dummy address for %s ` The `%s` will be replaced by the pk. However, the current codes will convert it to the uppercase if I call pk() function on the whole object or retrieve the pk from db directly, so the final dummy attribute will be `address: Dummy address for IRT-YOUMINET-CN` instead of `address: Dummy address for IRT-Youminet-CN`.
> > I can't find an easy way to find the original pk from the whole object.
> 
> There is none. The data simply isn't extracted, as PKs are case insensitive in every other usage, and therefore normalised to upper case during parsing of the initial object. Is it such an obstacle for you? Does the PK need to be repeated in the dummy text? After all, anyone who looks at the object will see the true PK already anyways.

Nvm, thanks.

Justin-APNIC commented:
> Left some notes inline. We also need to extend this to the source export runner, and the NRTM4 server. In those cases, we don't have client IPs, so our only option is to always enable it if configured.
> 
Ok, we need this because we have some internal clients for accessing the IRRD NRTM3 server, so it needs the original data.

> I also want to tweak the docs and settings names a bit, later.

No problem, thanks for looking into it.

Justin-APNIC committed:
Delete local test reports

Justin-APNIC committed:
Overwrite the auth attribute if there is a dummy one

Justin-APNIC committed:
lint

mxsasha commented:
I'm going to close this as it seems stale. Do reopen in the future if you want to work on it, I think the feature can make sense.

Justin-APNIC commented:
> I'm going to close this as it seems stale. Do reopen in the future if you want to work on it, I think the feature can make sense.

Hi, i think i have addressed all your comments above. Is there anything I can do for the mr? I still want to continue working on the feature. Thanks.

mxsasha committed:
cleanup after rebase

mxsasha commented:
> > I'm going to close this as it seems stale. Do reopen in the future if you want to work on it, I think the feature can make sense.
> 
> Hi, i think i have addressed all your comments above. Is there anything I can do for the mr? I still want to continue working on the feature. Thanks.

I think this one is still missing:

> We also need to extend this to the source export runner, and the NRTM4 server. In those cases, we don't have client IPs, so our only option is to always enable it if configured.

Justin-APNIC committed:
Implement dummy object feature in export runner

Justin-APNIC committed:
Introduce a new config for enabling dummy object export

Justin-APNIC committed:
Implement dummy object feature in nrtmv4 server

Justin-APNIC commented:
> I think this one is still missing:
> 
> > We also need to extend this to the source export runner, and the NRTM4 server. In those cases, we don't have client IPs, so our only option is to always enable it if configured.

Thanks for reopening the mr. This has been done.

mxsasha committed:
Lots of adjustments

mxsasha commented:
Thanks for those changes! I have just pushed a commit with a lot of smaller adjustments like style and naming tweaks. ~~And apparently broke a bunch of tests, which I will fix now.~~

I have also dropped two settings for now: the separate settings for dummifying exports, and the separate access list for undummified NRTMv3. My motivation is that NRTMv3 and exports are typically used together, and allowing inconsistency between them creates frustrating data inconsistencies, so we should not make that too easy. Also, the separation no longer exists in NRTMv4, on purpose, which is the protocol everyone should be migrating to anyways once it's ready.
